dask/highlevelgraph.py

```diff
@@ -1046,6 +1274,34 @@ def __dask_distributed_pack__(
         data: dict
             Packed high level graph layers
         """
+
+        # If pickle is disabled on the scheduler, all layers must
+        # be converted to `MaterializedLayer` objects before packing
+        materialize = not config.get("distributed.scheduler.pickle")
```

Comment:
Yeah - I guess it is possible for the config setting to be different on the scheduler and client, but I suspect this will cover most cases.

When the client tries to send pickled data to the scheduler, the scheduler [will throw an error](https://github.com/dask/distributed/blob/2e98ae97c9f1b7487abb309207f6a455e3c7ed37/distributed/protocol/core.py#L162).  Perhaps we should just update this error in distributed to clarify that the config must be the "pickle" config must be True on both the client **and** scheduler?

Reply:
> I suspect this will cover most cases

I think this is true for local clusters, but if there is a remote cluster somewhere with pickling disabled, I expect it to fail more often than it succeeds (simply because most users won't have a local config for this).

The error message and documentation may be enough, though I still think a better user experience would be to introspect something on the `client` (which is an arg here) to determine whether to pickle something. Happy to leave that discussion for a follow-up, however, since it's separate from API design.

Reply:
>The error message and documentation may be enough, though I still think a better user experience would be to introspect something on the client (which is an arg here) to determine whether to pickle something. Happy to leave that discussion for a follow-up, however, since it's separate from API design.

I agree with the user-experience concern. Even with a clear error message, it will probably be confusing to run into serialization errors when the external cluster has pickling disabled (however rare it may be for people to explicitly disable pickling). I don't personally have any bright ideas about how we can introspect this information on the `client`. Do you think we need to implement something new to make this information available - I'm hoping we dont need to do something annoying, like ping the scheduler with a pickled test packet.

Reply:
> Do you think we need to implement something new to make this information available - I'm hoping we dont need to do something annoying, like ping the scheduler with a pickled test packet.

Yeah, we probably would need to implement something new, I don't think this is possible today (at least, not without pickling a test packet as you suggest). But I don't think there is a fundamental reason it *couldn't* be available. But I'm happy to defer this discussion!